lib/core/proxy.c

```diff
@@ -183,28 +187,31 @@ static h2o_iovec_t build_request(h2o_req_t *req, int keepalive)
         buf.base[offset++] = '\r';
         buf.base[offset++] = '\n';
     }
-    FLATTEN_PREFIXED_VALUE("x-forwarded-proto: ", req->input.scheme->name, 0);
-    buf.base[offset++] = '\r';
-    buf.base[offset++] = '\n';
-    if (remote_addr_len != SIZE_MAX) {
-        FLATTEN_PREFIXED_VALUE("x-forwarded-for: ", xff_buf, remote_addr_len);
-        APPEND(remote_addr, remote_addr_len);
-    } else {
-        FLATTEN_PREFIXED_VALUE("x-forwarded-for: ", xff_buf, 0);
-    }
-    buf.base[offset++] = '\r';
-    buf.base[offset++] = '\n';
-    FLATTEN_PREFIXED_VALUE("via: ", via_buf, sizeof("1.1 ") - 1 + req->input.authority.len);
-    if (req->version < 0x200) {
-        buf.base[offset++] = '1';
-        buf.base[offset++] = '.';
-        buf.base[offset++] = '0' + (0x100 <= req->version && req->version <= 0x109 ? req->version - 0x100 : 0);
-    } else {
-        buf.base[offset++] = '2';
+    if (!is_websocket_handshake) {
+        FLATTEN_PREFIXED_VALUE("x-forwarded-proto: ", req->input.scheme->name, 0);
+        buf.base[offset++] = '\r';
+        buf.base[offset++] = '\n';
+        if (remote_addr_len != SIZE_MAX) {
+            FLATTEN_PREFIXED_VALUE("x-forwarded-for: ", xff_buf, remote_addr_len);
+            APPEND(remote_addr, remote_addr_len);
+        } else {
+            FLATTEN_PREFIXED_VALUE("x-forwarded-for: ", xff_buf, 0);
+        }
+        buf.base[offset++] = '\r';
+        buf.base[offset++] = '\n';
+        FLATTEN_PREFIXED_VALUE("via: ", via_buf, sizeof("1.1 ") - 1 + req->input.authority.len);
+        if (req->version < 0x200) {
+            buf.base[offset++] = '1';
+            buf.base[offset++] = '.';
+            buf.base[offset++] = '0' + (0x100 <= req->version && req->version <= 0x109 ? req->version - 0x100 : 0);
+        } else {
+            buf.base[offset++] = '2';
+        }
+        buf.base[offset++] = ' ';
+        APPEND(req->input.authority.base, req->input.authority.len);
+        APPEND_STRLIT("\r\n");
```

Comment:
please include all headers regardless of whether if the request is a websocket handshake.  Application servers may use the information provided by the headers (e.g. x-forwarded-for) to determine if it should allow opening a websocket connection.